### PR TITLE
Add support for non aligned SIMD for vec4

### PR DESCRIPTION
**Pull request by laurentcau**
It is used as default configuration for Visual Studio 64 bits compilation (needs Language Extension). 
code changes:
- add new qualifiers: unaligned_simd_highp unaligned_simd_mediump unaligned_simd_lowp
- add use_simd and replace is_aligned 
(code for ARM NEON is added but not tested)

### CODE REVIEW

**Review by christophe-lunarg — APPROVED** [review]
Thanks for contributing

File: glm/detail/qualifier.hpp [hunk 1]
```diff
@@ -173,17 +283,56 @@ namespace detail
...
+		typedef struct type {
+			float data[4];
+			GLM_DEFAULTED_DEFAULT_CTOR_QUALIFIER GLM_CONSTEXPR type() GLM_DEFAULT;
+			inline type(glm_f32vec4 v) { vst1q_f32(reinterpret_cast<float*>(data), v); }
+			inline operator glm_f32vec4() const { return vld1q_f32(reinterpret_cast<const float*>(data)); }
+		} type;
+	};
+
+
+	return ;
```
Comment (tycho) [inline glm/detail/qualifier.hpp:298]: This line shouldn't be here.

File: glm/detail/qualifier.hpp [hunk 2]
```diff
@@ -173,17 +283,56 @@ namespace detail
...
 
+	template<>
+	struct storage<4, int, false, true>
+	{
+		struct type
+		{
+			int data[4];
+			GLM_DEFAULTED_DEFAULT_CTOR_QUALIFIER GLM_CONSTEXPR type() GLM_DEFAULT;
+			type(glm_i32vec4 v) { vst1q_u32(data, v); }
+			operator glm_i32vec4() const { return vld1q_u32(data); }
```
Comment (tycho) [inline glm/detail/qualifier.hpp:314]: These should be `vst1q_s32` and `vld1q_s32`, since the vector is signed.

File: glm/detail/qualifier.hpp [hunk 3]
```diff
@@ -173,17 +283,56 @@ namespace detail
...
+
+	template<>
+	struct storage<4, unsigned int, false, true>
+	{
+		struct type
+		{
+			unsigned int data[4];
+			GLM_DEFAULTED_DEFAULT_CTOR_QUALIFIER GLM_CONSTEXPR type() GLM_DEFAULT;
+			type(glm_i32vec4 v) { vst1q_u32(data, v); }
+			operator glm_i32vec4() const { return vld1q_u32(data); }
```
Comment (tycho) [inline glm/detail/qualifier.hpp:332]: `glm_i32vec4` should be replaced with `glm_u32vec4` here.

**Review by tycho — COMMENTED** [review]
Some bugs in this contribution that I just ran into when trying to build targeting NEON.

**Comment by laurentcau** [comment]
Hi,

Feal free to fix the mistake. I have no way to test ARM version and glm validation process only support Windows Visual Studio build.
If you know an easy way to test this on PC, share it, please.
I have another big Pull Request that will surely break ARM version too. (add simd (aligned and unaligned) operation for vec3 and dvec3 by laurentcau · Pull Request #1164 · g-truc/glm · GitHub<https://github.com/g-truc/glm/pull/1164>)

Regards
Laurent

**Comment by tycho** [comment]
> Hi, Feal free to fix the mistake. I have no way to test ARM version and glm validation process only support Windows Visual Studio build. If you know an easy way to test this on PC, share it, please.

Well, I built with clang-cl targeting ARM64 on Visual Studio to find these issues. If you install clang-cl and the other ARM64 bits via the Visual Studio Installer, you should be able to test that. You can configure with CMake using either `-G "Visual Studio 16 2019" -A ARM64 -T ClangCL` or `-G "Visual Studio 16 2019" -A ARM64 -T LLVM_v142`, depending on which toolset is installed.

**Comment by laurentcau** [comment]
Thank you for the tip.
I was able to setup a build. It compiles fine BUT it doesn’t use the NEON path.
125>D:\repo\ext\glm\glm/gtc/../ext/../detail/setup.hpp(948,11): warning : "GLM: version 0.9.9.9" [-W#pragma-messages]
125>D:\repo\ext\glm\glm/gtc/../ext/../detail/setup.hpp(960,11): warning : GLM: C++ 14 with extensions [-W#pragma-messages]
125>D:\repo\ext\glm\glm/gtc/../ext/../detail/setup.hpp(991,11): warning : GLM: Clang compiler detected [-W#pragma-messages]
125>D:\repo\ext\glm\glm/gtc/../ext/../detail/setup.hpp(1061,11): warning : GLM: Unknown build target [-W#pragma-messages]
125>D:\repo\ext\glm\glm/gtc/../ext/../detail/setup.hpp(1074,11): warning : GLM: Windows platform detected [-W#pragma-messages]
125>D:\repo\ext\glm\glm/gtc/../ext/../detail/setup.hpp(1096,11): warning : GLM: GLM_FORCE_SWIZZLE is defined, swizzling operators enabled. [-W#pragma-messages]
125>D:\repo\ext\glm\glm/gtc/../ext/../detail/setup.hpp(1107,11): warning : GLM: GLM_FORCE_SIZE_T_LENGTH is undefined. .length() returns a glm::length_t, a typedef of int following GLSL. [-W#pragma-messages]
125>D:\repo\ext\glm\glm/gtc/../ext/../detail/setup.hpp(1113,11): warning : GLM: GLM_FORCE_UNRESTRICTED_GENTYPE is undefined. Follows strictly GLSL on valid function genTypes. [-W#pragma-messages]
125>D:\repo\ext\glm\glm/gtc/../ext/../detail/setup.hpp(1119,11): warning : GLM: GLM_FORCE_SILENT_WARNINGS is undefined. Shows C++ warnings from using C++ language extensions. [-W#pragma-messages]
125>D:\repo\ext\glm\glm/gtc/../ext/../detail/setup.hpp(1146,11): warning : GLM: GLM_FORCE_DEPTH_ZERO_TO_ONE is undefined. Using negative one to one depth clip space. [-W#pragma-messages]
125>D:\repo\ext\glm\glm/gtc/../ext/../detail/setup.hpp(1152,11): warning : GLM: GLM_FORCE_LEFT_HANDED is undefined. Using right handed coordinate system. [-W#pragma-messages]

Is there any option I should add to cmake to use it ?

**Comment by tycho** [comment]
Oh yeah, my branch has patches to make it detect ARM Windows (`_M_ARM` and `_M_ARM64`). You can add -DGLM_FORCE_NEON=1 and it should configure properly.

**Comment by christophe-lunarg** [comment]
Hi Laurent,

I revert the commit because it cause too much issues. I am working on fixing the repository C.I. with https://github.com/g-truc/glm/pull/1167 which I hope will help avoiding such problem in the future...

Thanks!

**Pull request merged**